zrml/swaps/runtime-api/src/lib.rs

```diff
@@ -19,16 +19,20 @@
 #![doc = include_str!("../README.md")]
 #![cfg_attr(not(feature = "std"), no_std)]
 
-use parity_scale_codec::{Codec, MaxEncodedLen};
-use sp_runtime::traits::{MaybeDisplay, MaybeFromStr};
-use zeitgeist_primitives::types::{Asset, SerdeWrapper};
+use parity_scale_codec::{Codec, Decode, MaxEncodedLen};
+use sp_runtime::{
+    traits::{MaybeDisplay, MaybeFromStr},
+    DispatchError,
+};
+use zeitgeist_primitives::types::{Asset, Pool, SerdeWrapper};
 
 sp_api::decl_runtime_apis! {
     pub trait SwapsApi<PoolId, AccountId, Balance, MarketId> where
         PoolId: Codec,
         AccountId: Codec,
-        Balance: Codec + MaybeDisplay + MaybeFromStr + MaxEncodedLen,
-        MarketId: Codec + MaxEncodedLen,
+        Balance: Codec + MaybeDisplay + MaybeFromStr + MaxEncodedLen + Decode,
+        MarketId: Codec + MaxEncodedLen + Decode,
```

Comment:
No need to ask for these to implement `Decode` if `Pool<Balance, MarketId>` is required to implement `Decode`.